modding-tutorials/harmony-basics.md

```diff
@@ -0,0 +1,181 @@
+- Parameter names starting with three underscores, for example `___someField`, can be used to read and write (with 'ref') private fields on the instance that has the same name (minus the underscores)
+- Patches can define only those parameters they want to access (no need to define all)
+- Patch parameters must use the exact same name and type as the original method (object is ok too)
+- Patches can either get parameters normally or by declaring any parameter ref (for manipulation)
+- To allow patch reusing, one can inject the original method by using a parameter named `__originalMethod`
+
+{% hint style="info" %} Using three underscores `___likeSo` to read variables is quite usefull {% endhint %}
+
+Transpilers have some other optional parameters:
+
+- A parameter of type ILGenerator that will be set to the current IL code generator
+- A parameter of type MethodBase that will be set to the current original method being patched
+- They must contain one parameter of type `IEnumerable<CodeInstruction>` that will be used to pass the IL codes to it
+
+
+{% hint style="info" %} There are other ways to combine attributes and structure your patches. Try checking [here](https://harmony.pardeike.net/articles/annotations.html#combining-annotations) if your mod starts getting complicated. {% endhint %}
+
+---
+### Applying your patches
+To get your code actually working Harmony needs to be told to do all the work you have set up for it. One place to do that from is Start method of your Mod.
+
+To do this make sure your file is already `using HarmonyLib;`
+
+Then you need to create a new instance of Harmony with an id for your set of patches. A common structure for an id is "com.company.project.product" or for us "com.User.Project.Feature"
+
+Then we tell the Harmony instance to patch everything it can find in our code. In older versions you need to tell Harmony to look for the currently running game with `Assembly.GetExecutingAssembly()` to try and patch it but these days it will look there by default.
+
+```csharp
+public void Start()
+    {
+        MyInput.Keybinds.Add("Alt", new Keybind("alt", KeyCode.LeftAlt, KeyCode.RightAlt));
+
+        harmonyInstance = new Harmony("com.Soggylithe.IgnoreBuildCollision");
+        harmonyInstance.PatchAll(Assembly.GetExecutingAssembly());
+
+        Debug.Log("Place Anywhere successfully loaded! - Default-AltKey");
+    }
+```
+
+Remember, anything you do to load your mod you need to undo when you unload it to prevent strange errors and to keep the client stable. This also means you need to tell your Harmony instance to unpatch the things you changed.
+
+```csharp
+public void OnModUnload()
+    {
+        MyInput.Keybinds.Remove("Alt");
+
+        harmonyInstance.UnpatchAll();
+        Destroy(gameObject);
+        Debug.Log("Mod Brine has been unloaded!");
+    }
+```
+
+In older versions you needed to `Destroy()` the mod object when unloading but that is also handled gracefully these days.
+
+## Traversing
```

Comment:
If you want, you can add your name at the bottom of the tutorial, but you don't need to :)